pints/tests/test_mcmc_haario_ac.py

```diff
@@ -0,0 +1,198 @@
+        mcmc = pints.HaarioACMCMC(x0)
+
+        # One round of ask-tell must have been run
+        self.assertRaisesRegex(
+            RuntimeError, 'already running', mcmc.replace, x0, 1)
+
+        mcmc.ask()
+
+        # One round of ask-tell must have been run
+        self.assertRaises(RuntimeError, mcmc.replace, x0, 1)
+
+        mcmc.tell(0.5)
+        mcmc.replace([1, 2, 3], 10)
+        mcmc.replace([1, 2, 3], 10)
+
+        # New position must have correct size
+        self.assertRaisesRegex(
+            ValueError, '`current` has the wrong dimensions',
+            mcmc.replace, [1, 2], 1)
+
+        # Proposal can be changed too
+        mcmc.ask()
+        mcmc.replace([1, 2, 3], 10, [3, 4, 5])
+
+        # New proposal must have correct size
+        self.assertRaisesRegex(
+            ValueError, '`proposed` has the wrong dimensions',
+            mcmc.replace, [1, 2, 3], 3, [3, 4])
+
+    def test_flow(self):
+
+        # Test initial proposal is first point
+        x0 = self.real_parameters
+        mcmc = pints.HaarioACMCMC(x0)
+        self.assertTrue(mcmc.ask() is mcmc._x0)
+
+        # Double initialisation
+        mcmc = pints.HaarioACMCMC(x0)
+        mcmc.ask()
+        self.assertRaises(RuntimeError, mcmc._initialise)
+
+        # Tell without ask
+        mcmc = pints.HaarioACMCMC(x0)
+        self.assertRaises(RuntimeError, mcmc.tell, 0)
+
+        # Repeated asks should return same point
+        mcmc = pints.HaarioACMCMC(x0)
+        # Get into accepting state
+        mcmc.set_initial_phase(False)
+        for i in range(100):
+            mcmc.tell(self.log_posterior(mcmc.ask()))
+        x = mcmc.ask()
+        for i in range(10):
+            self.assertTrue(x is mcmc.ask())
+
+        # Repeated tells should fail
+        mcmc.tell(1)
+        self.assertRaises(RuntimeError, mcmc.tell, 1)
+
+        # Bad starting point
+        mcmc = pints.HaarioACMCMC(x0)
+        mcmc.ask()
+        self.assertRaises(ValueError, mcmc.tell, float('-inf'))
+
+    def test_options(self):
+
+        # Test setting acceptance rate
+        x0 = self.real_parameters
+        mcmc = pints.HaarioACMCMC(x0)
+        self.assertNotEqual(mcmc.target_acceptance_rate(), 0.5)
+        mcmc.set_target_acceptance_rate(0.5)
+        self.assertEqual(mcmc.target_acceptance_rate(), 0.5)
+        mcmc.set_target_acceptance_rate(1)
+        self.assertRaises(ValueError, mcmc.set_target_acceptance_rate, 0)
+        self.assertRaises(ValueError, mcmc.set_target_acceptance_rate, -1e-6)
+        self.assertRaises(ValueError, mcmc.set_target_acceptance_rate, 1.00001)
+
+        # test hyperparameter setters and getters
+        self.assertEqual(mcmc.n_hyper_parameters(), 1)
+        self.assertRaises(ValueError, mcmc.set_hyper_parameters, [-0.1])
+        mcmc.set_hyper_parameters([0.3])
+        self.assertEqual(mcmc.eta(), 0.3)
+
+        self.assertEqual(mcmc.name(), 'Haario adaptive covariance MCMC')
+
+    def test_logging(self):
```

Comment:
Again, might have to go into test for base class

Reply:
Ok. At the moment, it tests that the name of each method is printed though...

Reply:
Good point!
I'm a bit conflicted, but I think the cleanest thing to do here is to think of the base class methods _using_ the inheriting methods. So we can maybe check that the log for 2 different methods has 2 different names in it, but this is a check of the base class, not the inheriting classes.
The unit tests for the inheriting classes, meanwhile, should check that their name() method works

Reply:
Yep, that's a good idea -- have changed this now so that the daughter class tests check for name and the abstract base class checks everything else (iterations, accept and so on).